colexec: audit BCE of Gets and Sets and add assertions in multiple places

yuzefovich opened the pull request:
**coldata: add BCE in Copy in some cases and in AppendSlice of decimals**

This commit audits the usage of slices in the `vec_tmpl` template and
improves `Copy` method with present selection vector for BCEs to occur
on sliceable types. This required introduction a separate path for
sliceable and non-sliceable types (which is a bit confusing, but the
performance improvement of 30-60% is worth it). Also, early bounds
checks are added to `AppendSlice` of decimals.

Additionally, this commit puts `gcassert:bce` in places where we expect
BCEs to occur. The caveat here that BCE works only on vectors that are
"sliceable", so this commit introduces a `Sliceable` method onto the
overload base struct.

Release note: None

**colconv: add early bounds checks and BCE assertions**

This commit reviews `vec_to_datum` template and introduces early bounds
checks where possible. Additionally, BCE assertions are added.

Release note: None

**colexecagg: enforce BCEs where possible**

This commit audits all aggregate functions to add BCE assertions where
we expect them to occur. The following issues were fixed:
- `bool_and_agg` missing the capture of `col`
- `default_agg` missing the capture and slicing of `groups`.

Release note: None

**colexec: improve hash utility methods with early bounds checks**

This commit adds early bounds checks to several hash utility methods
(enforced by the BCE assertions). This required a minor change to the
hash aggregator to avoid a scenario of aggregating over zero buffered
tuples.

Release note: None

**colexec: add early bounds checks in columnarizer and hashtable**

Release note: None

**colexec: audit most templates to add BCE assertions**

This commit reviews most templates in the colexec package to add BCE
assertions where we expect them to occur. The following problems were
fixed:
- `distinct_tmpl` had BCE in the wrong place so that it wasn't working
- `proj_const` missing the capture of `col`.

New early bounds checks are now applied in:
- multiple slices in window functions
- `rowstovec` template
- `hash_aggregator` template
- multiple functions in `hashtable` template
- `default_cmp` template.

Release note: None

**colexec: audit the hash joiner to get BCE**

This commit refactors the hash joiner code to allow for bounds checks
eliminations to occur. Previously, we tried to get them working, but
they didn't because we were missing the capture of the variables, so the
compiler couldn't prove that BCE was safe. This commit adds the early
bounds checks in almost all places in the template file (except for two
spots where I couldn't figure out how to get BCE to occur) which gives
us a modest performance improvement. This, however, required to adjust
the logic slightly.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57854)
<!-- Reviewable:end -->

jordanlewis reviewed (commented):
Great stuff! Thanks for doing this. I think you could add some more `//gcassert:bce` in all the places you added early bounds checks, right?

> Reviewed 31 of 63 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-:-MOUGIKIBxelOxbDmeTV:bhakc9z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colconv/vec_to_datum.eg.go, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUEmyeEl0EBRZdMJgP:-MOUEmyeEl0EBRZdMJgQ:b1ho48p) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colconv/vec_to_datum.eg.go#L217)):*
> ```Go
> 	if col.MaybeHasNulls() {
> 		nulls := col.Nulls()
> 		_ = converted[length-1]
> ```

Does this actually do anything? I think you would need to change `converted[destIdx]` to `converted[idx]`, but maybe not?
___
*[pkg/sql/colconv/vec_to_datum.eg.go, line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUEUL_7gdtlS4TAS0a:-MOUEUL_7gdtlS4TAS0b:b-xhmmfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colconv/vec_to_datum.eg.go#L228)):*
> ```Go
> 				for idx = 0; idx < length; idx++ {
> 					destIdx = idx
> 					srcIdx = sel[idx]
> ```

Can these lines get a `//gcassert:bce` or no?
___
*[pkg/sql/colconv/vec_to_datum.eg.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUEteC36iJxa_I6oOA:-MOUEteC36iJxa_I6oOB:b-horza2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colconv/vec_to_datum.eg.go#L230)):*
> ```Go
> 					srcIdx = sel[idx]
> 					if nulls.NullAt(srcIdx) {
> 						converted[destIdx] = tree.DNull
> ```

`//gcassert:bce`?
___
*[pkg/sql/colconv/vec_to_datum.eg.go, line 233 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUEv8452ixcXzUFOFC:-MOUEv8452ixcXzUFOFD:b-horza2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colconv/vec_to_datum.eg.go#L233)):*
> ```Go
> 						continue
> 					}
> 					converted[destIdx] = da.NewDName(tree.DString(bytes.Get(srcIdx)))
> ```

`//gcassert:bce`?
___
*[pkg/sql/colexec/distinct.eg.go, line 910 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFDIj8ZNr4-gpM7Sd:-MOUFDIj8ZNr4-gpM7Se:b-horza2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/distinct.eg.go#L910)):*
> ```Go
> 							if !lastValNull {
> 								// The current value is null while the previous was not.
> 								outputCol[outputIdx] = true
> ```

`//gcassert:bce`?
___
*[pkg/sql/colexec/distinct.eg.go, line 916 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFECW14a50bNHsFr2:-MOUFECW14a50bNHsFr3:b-horza2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/distinct.eg.go#L916)):*
> ```Go
> 							if lastValNull {
> 								// The previous value was null while the current is not.
> 								outputCol[outputIdx] = true
> ```

`//gcassert:bce`?
___
*[pkg/sql/colexec/hash_aggregator_tmpl.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFcM7DN-sIZJuCWCB:-MOUFcM7DN-sIZJuCWCC:b-horza2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/hash_aggregator_tmpl.go#L42)):*
> ```Go
> 		// eqChainsID that we're currently building in which eqChainsID
> 		// indicates that the current tuple is the head of its equality chain.
> 		headID := op.ht.probeScratch.headID[i]
> ```

`//gcassert:bce`?
___
*[pkg/sql/colexec/hash_aggregator_tmpl.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFW1g6rm5txCojkwt:-MOUFW1g6rm5txCojkwu:b-24mfjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/hash_aggregator_tmpl.go#L49)):*
> ```Go
> 			op.ht.probeScratch.hashBuffer[eqChainsCount] = op.ht.probeScratch.hashBuffer[i]
> 			if useSel {
> 				sel[eqChainsCount] = sel[i]
> ```

`//gcassert:bce`? (probably you'd need to break up the two `sel` accesses)
___
*[pkg/sql/colexec/hashtable_tmpl.go, line 506 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFiBpAP-gpf9WZbPI:-MOUFiBpAP-gpf9WZbPJ:b-horza2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/hashtable_tmpl.go#L506)):*
> ```Go
> 	copy(visited, zeroBoolColumn)
> 	for i := 0; i < batchLength; i++ {
> 		if headID := ht.probeScratch.headID[i]; headID != 0 {
> ```

`//gcassert:bce`?
___
*[pkg/sql/colexec/rank_tmpl.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFrgz6qPy7lMfmgFY:-MOUFrgz6qPy7lMfmgFZ:b-b09763) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/rank_tmpl.go#L181)):*
> ```Go
> 		// {{end}}
> 		_ = peersCol[n-1]
> 		_ = rankCol[n-1]
> ```

`//gcassert:bce` for the accesses that are changed by these?
___
*[pkg/sql/colexec/relative_rank_tmpl.go, line 537 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUG0K57iugJGWAj8FI:-MOUG0K57iugJGWAj8FJ:bium1e5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/relative_rank_tmpl.go#L537)):*
> ```Go
> 				// computed).
> 				// {{if .HasPartition}}
> 				if partitionCol[i] {
> ```

`//gcassert:bce`? (ditto for other ones checked above)
___
*[pkg/sql/colexec/row_number_tmpl.go, line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUG4999mbB68mhyxRF:-MOUG4999mbB68mhyxRG:bsx31jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/row_number_tmpl.go#L118)):*
> ```Go
> 	} else {
> 		// {{if .HasPartition}}
> 		_ = partitionCol[n-1]
> ```

`//gcassert:bce` for later accesses?
___
*[pkg/sql/colexec/row_number_tmpl.go, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUG3vV2qtIFMlMXMx7:-MOUG3vV2qtIFMlMXMx8:bsx31jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/row_number_tmpl.go#L120)):*
> ```Go
> 		_ = partitionCol[n-1]
> 		// {{end}}
> 		_ = rowNumberCol[n-1]
> ```

`//gcassert:bce` for later accesses?
___
*[pkg/sql/colexec/window_peer_grouper_tmpl.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUGAf64JTQ--XjeOZd:-MOUGAf7AxFMeI62wuG2:bsx31jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/window_peer_grouper_tmpl.go#L181)):*
> ```Go
> 		_ = peersCol[n-1]
> 		_ = partitionCol[n-1]
> 		_ = p.distinctCol[n-1]
> ```

`//gcassert:bce` for later accesses?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
As mentioned, at first, I thought I would add the assertions only in the "critical" places (without a clear definition of what "critical" would mean), but I think you're right - given we make all this effort to have BCE, we might as well assert in all cases where we expect them.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-:-MOY-4J47DMRBIml8VFP:b-l0lnya)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/colconv/vec_to_datum.eg.go, line 217 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUEmyeEl0EBRZdMJgP:-MOYKnN04Ls_oPaUeOVM:b-bejf3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colconv/vec_to_datum.eg.go#L217)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Does this actually do anything? I think you would need to change `converted[destIdx]` to `converted[idx]`, but maybe not?
</blockquote></details>

It does appear to work according to the linter and go compiler.
___
*[pkg/sql/colconv/vec_to_datum.eg.go, line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUEUL_7gdtlS4TAS0a:-MOYKrdT8a7ZI2O1oTep:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colconv/vec_to_datum.eg.go#L228)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Can these lines get a `//gcassert:bce` or no?
</blockquote></details>

Done.
___
*[pkg/sql/colconv/vec_to_datum.eg.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUEteC36iJxa_I6oOA:-MOYKu6q8rTSMVwmPr3x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colconv/vec_to_datum.eg.go#L230)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce`?
</blockquote></details>

Done.
___
*[pkg/sql/colconv/vec_to_datum.eg.go, line 233 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUEv8452ixcXzUFOFC:-MOYKuWG-x4s2KAP1A0t:b-80xpnf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colconv/vec_to_datum.eg.go#L233)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce`?
</blockquote></details>

This one is confusing, left a TODO.
___
*[pkg/sql/colexec/distinct.eg.go, line 910 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFDIj8ZNr4-gpM7Sd:-MOYOivT0DrytpgiKGkS:b3wgrp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/distinct.eg.go#L910)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce`?
</blockquote></details>

Here, we can't add the assertions right now because the "new" execgen doesn't support omitting some of the code based on a conditional. We end up having something like
```
if false {
  //gcassert:bce
}
outputCol[outputIdx] = true
...
```
when there is a selection vector present. We'll need to think of a way to go around it, left a TODO.
___
*[pkg/sql/colexec/distinct.eg.go, line 916 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFECW14a50bNHsFr2:-MOYPZOd7WQwJq-LdrxM:b-816h6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/distinct.eg.go#L916)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce`?
</blockquote></details>

Same.
___
*[pkg/sql/colexec/hash_aggregator_tmpl.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFcM7DN-sIZJuCWCB:-MOYQKA02AAMgvGJa0aV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/hash_aggregator_tmpl.go#L42)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce`?
</blockquote></details>

Done.
___
*[pkg/sql/colexec/hash_aggregator_tmpl.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFW1g6rm5txCojkwt:-MOYQKja0LqnY24dy67I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/hash_aggregator_tmpl.go#L49)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce`? (probably you'd need to break up the two `sel` accesses)
</blockquote></details>

Done.
___
*[pkg/sql/colexec/hashtable_tmpl.go, line 506 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFiBpAP-gpf9WZbPI:-MOY_sxN6oSMlf08FyS2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/hashtable_tmpl.go#L506)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce`?
</blockquote></details>

Done.
___
*[pkg/sql/colexec/rank_tmpl.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUFrgz6qPy7lMfmgFY:-MOZDinY-seU1n3FdgEe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/rank_tmpl.go#L181)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce` for the accesses that are changed by these?
</blockquote></details>

Done.
___
*[pkg/sql/colexec/relative_rank_tmpl.go, line 537 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUG0K57iugJGWAj8FI:-MOZEh2i5yDpsMjBzv_W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/relative_rank_tmpl.go#L537)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce`? (ditto for other ones checked above)
</blockquote></details>

Done.
___
*[pkg/sql/colexec/row_number_tmpl.go, line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUG4999mbB68mhyxRF:-MOZEzKDE7ffX6hhFuQR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/row_number_tmpl.go#L118)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce` for later accesses?
</blockquote></details>

Done.
___
*[pkg/sql/colexec/row_number_tmpl.go, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUG3vV2qtIFMlMXMx7:-MOZEzgnET2UKcOY4p0t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/row_number_tmpl.go#L120)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce` for later accesses?
</blockquote></details>

Done.
___
*[pkg/sql/colexec/window_peer_grouper_tmpl.go, line 181 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUGAf64JTQ--XjeOZd:-MOZH9TME6dldOimE2A7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colexec/window_peer_grouper_tmpl.go#L181)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`//gcassert:bce` for later accesses?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Ran hash joiner benchmarks [here](https://gist.github.com/yuzefovich/87ae53e27508b18d2ea10eb3a50c6440) (since I fought the most to get as many BCE to occur as possible).

yuzefovich commented:
Added some BCE in `Vec.Copy` for simple types - very nice wins [there](https://gist.github.com/yuzefovich/d55cca70f46f4ac0aa27b5ef2b2dd0bd):
```
Copy/int/CopySimple/NullProbability=0.0-24       66.5GB/s ± 2%   67.6GB/s ± 1%   +1.73%  (p=0.000 n=10+9)
Copy/int/CopyWithSel/NullProbability=0.0-24      6.87GB/s ± 1%  11.23GB/s ± 1%  +63.43%  (p=0.000 n=10+10)
Copy/int/CopySimple/NullProbability=0.2-24       43.6GB/s ± 1%   44.7GB/s ± 2%   +2.38%  (p=0.000 n=10+10)
Copy/int/CopyWithSel/NullProbability=0.2-24      2.82GB/s ± 2%   3.99GB/s ± 1%  +41.19%  (p=0.000 n=10+10)
```

jordanlewis commented:
Nice... addictive isn't it? Let me know when it is RFAL again.

yuzefovich commented:
Yeah, it is quite addictive lol.

I think it is RFAL - I reviewed all template files (I initially forgot about `vec_tmpl`) and addressed all simple TODOs.

yuzefovich commented:
Oh, I guess I can merge BCE assertions for `execgen.SET` from the other PR (without trying to make `Set` inlinable).

yuzefovich reviewed (commented):
Alright, merged the changes from the other PR, RFAL.

The only template I didn't audit is the merge joiner ^_^.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-:-MOZmytZ4eon8hfGite8:b-4tggva)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/colconv/vec_to_datum.eg.go, line 233 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOUEv8452ixcXzUFOFC:-MOZmo6qCt2XxKb9iE0j:bm76s8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/3843a1fe5242e62f3df081c1f57e1876252f6387/pkg/sql/colconv/vec_to_datum.eg.go#L233)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This one is confusing, left a TODO.
</blockquote></details>

Kept the TODO for the harder general case only.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Addressed the remaining TODO in `vec_to_datum`, added another commit with BCEs, and ran the benchmarks of the whole package [here](https://gist.github.com/yuzefovich/e6babc719267a57c434c623c1f0bd021). The ordered distinct seems to have gotten slower, so I'll take a closer look at it.

yuzefovich commented:
Fixed the largest regressions on the ordered distinct and `bool_` aggregates (benchmarks are [here](https://gist.github.com/yuzefovich/fed8cdc3790d7ef3c45724751e50df27)).

I think it is RFAL, and I'll stop looking for more BCE opportunities (at least for now :) ).

jordanlewis reviewed (approved):
Wait, why would anything have gotten slower at all? I don't understand that.

Anyway, this is a big PR that I think accidentally got too big, I may have encouraged that by accident too. I don't want to ask you to break it up because that would take a lot of your time, but a gentle plea to try to keep logic changes out of mechanical PRs, and in general to keep PRs smaller.

> Reviewed 2 of 60 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-:-MOntk8F6FB0SgwbQuDb:b1pxroi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/col/coldata/vec_tmpl.go, line 128 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOnIr2ZEd5Cc2zneYz8:-MOnIr2ZEd5Cc2zneYz9:be5jh46) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/col/coldata/vec_tmpl.go#L128)):*
> ```Go
> 				//gcassert:bce
> 				execgen.SET(toCol, i, v)
> 				// {{else}}
> ```

I don't like that the loop is half in 0-coords and half in destIdx coords... so confusing! What about this?

```
end := n+args.DestIdx
for i := args.DestIdx; i < end; i++ {
}
```
___
*[pkg/col/coldata/vec_tmpl.go, line 152 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOnJ6Lw0ZQBNB4RZ9e0:-MOnJ6Lw0ZQBNB4RZ9e1:b-z3p5ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/col/coldata/vec_tmpl.go#L152)):*
> ```Go
> 		// {{if .Sliceable}}
> 		//gcassert:bce
> 		execgen.SET(toCol, i, v)
> ```

ditto above
___
*[pkg/sql/colexec/columnarizer.go, line 196 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOnt6hZ9ImhI1qEb5py:-MOnt6h_Ea-SoqWWPElZ:b-8w7o32) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/sql/colexec/columnarizer.go#L196)):*
> ```Go
> 	outputRows := c.buffered[:nRows]
> 	for idx, ct := range c.typs {
> 		//gcassert:bce
> ```

What does this one do? Probably nothing? We should make `gcassert:bce` fail is there is no bounds to be checked at all perhaps
___
*[pkg/sql/colexec/hash_aggregator.go, line 200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOnudX56CHmH8juKrNt:-MOnudX56CHmH8juKrNu:bixv2bz) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/sql/colexec/hash_aggregator.go#L200)):*
> ```Go
> 			n := op.bufferingState.pendingBatch.Length()
> 			if n == 0 {
> 				if op.bufferingState.tuples.Length() == 0 {
> ```

Why did this stuff change? I don't see anything in the commit message about it.
___
*[pkg/sql/colexec/hashjoiner_tmpl.go, line 45 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOnwqto5Db4YhXH5zlc:-MOnwqto5Db4YhXH5zld:b-h4ne3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/sql/colexec/hashjoiner_tmpl.go#L45)):*
> ```Go
> 			// For some reason, BCE doesn't occur for probeRowUnmatched slice.
> 			// TODO(yuzefovich): figure it out.
> 			hj.probeState.probeRowUnmatched[nResults] = rowUnmatched
> ```

It is because `hj` is on the heap, and we have a pointer to it. The compiler can't prove that nobody else has a pointer to `hj` who could have edited the value of its `probeState.probeRowUnmatched` field. To fix this, it's the same as the capturing thing - you'd have to grab the slice in a variable before the loop (Of course we'd have to prove ourselves that nobody is modifying the slice, but we know that nobody is here.)
___
*[pkg/sql/colexec/hashjoiner_tmpl.go, line 115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOo2LWl3odHg8u753ny:-MOo2LWl3odHg8u753nz:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/sql/colexec/hashjoiner_tmpl.go#L115)):*
> ```Go
> 			// For some reason, BCE doesn't occur for buildIdx slice.
> 			// TODO(yuzefovich): figure it out.
> 			hj.probeState.buildIdx[nResults] = int(currentID - 1)
> ```

same as above
___
*[pkg/sql/colexec/hashjoiner_tmpl.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOo2SD3-1WzEuDBR77V:-MOo2SD3-1WzEuDBR77W:br4jd2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/sql/colexec/hashjoiner_tmpl.go#L124)):*
> ```Go
> 		}
> 
> 		if nResults == maxResults {
> ```

These changes are too big for what is supposed to be a mechanical transformation - it's not ideal to have logic changes like this buried in such a big commit (for one it'll be hard to understand what of this diff made a problem if we turn one up later).


<!-- Sent from Reviewable.io -->

yuzefovich committed:
coldata: add BCE in Copy in some cases and in AppendSlice of decimals

This commit audits the usage of slices in the `vec_tmpl` template and
improves `Copy` method with present selection vector for BCEs to occur
on sliceable types. This required introduction a separate path for
sliceable and non-sliceable types (which is a bit confusing, but the
performance improvement of 30-60% is worth it). Also, early bounds
checks are added to `AppendSlice` of decimals.

Additionally, this commit puts `gcassert:bce` in places where we expect
BCEs to occur. The caveat here that BCE works only on vectors that are
"sliceable", so this commit introduces a `Sliceable` method onto the
overload base struct.

Release note: None

yuzefovich committed:
colconv: add early bounds checks and BCE assertions

This commit reviews `vec_to_datum` template and introduces early bounds
checks where possible. Additionally, BCE assertions are added.

Release note: None

yuzefovich committed:
colexecagg: enforce BCEs where possible

This commit audits all aggregate functions to add BCE assertions where
we expect them to occur. The following issues were fixed:
- `bool_and_agg` missing the capture of `col`
- `default_agg` missing the capture and slicing of `groups`.

Release note: None

yuzefovich committed:
colexec: audit most templates to add BCE assertions

This commit reviews most templates in the colexec package to add BCE
assertions where we expect them to occur. The following problems were
fixed:
- `distinct_tmpl` had BCE in the wrong place so that it wasn't working
- `proj_const` missing the capture of `col`.

New early bounds checks are now applied in:
- multiple slices in window functions
- `rowstovec` template
- `hash_aggregator` template
- multiple functions in `hashtable` template
- `default_cmp` template.

Release note: None

yuzefovich committed:
colexec: audit the hash joiner to get BCE

This commit refactors the hash joiner code to allow for bounds checks
eliminations to occur. Previously, we tried to get them working, but
they didn't because we were missing the capture of the variables, so the
compiler couldn't prove that BCE was safe. This commit adds the early
bounds checks in almost all places in the template file (except for two
spots where I couldn't figure out how to get BCE to occur) which gives
us a modest performance improvement. This, however, required to adjust
the logic slightly.

Release note: None

yuzefovich committed:
colexec: improve hash utility methods with early bounds checks

This commit adds early bounds checks to several hash utility methods
(enforced by the BCE assertions). This required a minor change to the
hash aggregator to avoid a scenario of aggregating over zero buffered
tuples.

Release note: None

yuzefovich committed:
colexec: add early bounds checks in columnarizer and hashtable

Release note: None

yuzefovich reviewed (commented):
I'd think that most of regressions are noise, but there were a couple on which - although BCE was occurring - the performance decreased (`distinct_tmpl` and `bool_and_or_agg`). I'll kick off another nightly run to get the final numbers.

I agree that the PR, and especially the first commit, has gotten too big - I think it started out of reasonable size, but over time it grew like a snowball since I wanted to audit all templates, and then I wanted to look into the hash table, and then in the hash joiner, etc, lol. I've broken it up into more reasonable pieces, and I should have done it right away.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-:-MOoOT53E5a6rgkztzFi:b-e7vuiu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/col/coldata/vec_tmpl.go, line 128 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOnIr2ZEd5Cc2zneYz8:-MOo7RcF3YuINF9w5UJK:bca5g9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/col/coldata/vec_tmpl.go#L128)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I don't like that the loop is half in 0-coords and half in destIdx coords... so confusing! What about this?

```
end := n+args.DestIdx
for i := args.DestIdx; i < end; i++ {
}
```
</blockquote></details>

I don't see a way around having this confusing logic in place. The difficulty is that we have a single variable `i` that is used to access `sel` and `toCol` slices, and in order to get BCE, we adjust those slices according to `SrcStartIdx` and `DestIdx`, respectively. However, not all `toCol` vectors are sliceable, so we need to have separate logic for that.

It's possible that there is something I'm not seeing, but I spent significant time thinking about it when working on it originally, didn't see a way to go around it and don't see it now :/

Added the comments at least.
___
*[pkg/col/coldata/vec_tmpl.go, line 152 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOnJ6Lw0ZQBNB4RZ9e0:-MOo8mGV6JQmcjXRWON7:b-ya81ng) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/col/coldata/vec_tmpl.go#L152)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

ditto above
</blockquote></details>

Same as above
___
*[pkg/sql/colexec/columnarizer.go, line 196 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOnt6hZ9ImhI1qEb5py:-MOo9_Hs8PveJSjoKeKM:b-401cdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/sql/colexec/columnarizer.go#L196)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What does this one do? Probably nothing? We should make `gcassert:bce` fail is there is no bounds to be checked at all perhaps
</blockquote></details>

Yeah, this was a left over, removed (this assertion prompted me to pull out `outputRows` outside of the loop).
___
*[pkg/sql/colexec/hash_aggregator.go, line 200 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOnudX56CHmH8juKrNt:-MOoHTtG8AoLzFl-Lher:b-150r50) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/sql/colexec/hash_aggregator.go#L200)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Why did this stuff change? I don't see anything in the commit message about it.
</blockquote></details>

This was necessary to avoid attempting to compute the hash on the batch of zero length since we now have early bounds checks like `_ = buckets[nKeys - 1]`, and on zero-length batches we'd get the index out of bounds error.

Although the conditional here might seem complicated, it is relatively trivial, and the code would benefit from having it here in the first place. Added some comments too.
___
*[pkg/sql/colexec/hashjoiner_tmpl.go, line 45 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOnwqto5Db4YhXH5zlc:-MOo90n7BkjXMpVMrPVY:b-dhoe5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/sql/colexec/hashjoiner_tmpl.go#L45)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It is because `hj` is on the heap, and we have a pointer to it. The compiler can't prove that nobody else has a pointer to `hj` who could have edited the value of its `probeState.probeRowUnmatched` field. To fix this, it's the same as the capturing thing - you'd have to grab the slice in a variable before the loop (Of course we'd have to prove ourselves that nobody is modifying the slice, but we know that nobody is here.)
</blockquote></details>

I tried many variants of the same approach I took with `buildIdx` and `probeIdx`, and none worked. I guess my TODO could have been more descriptive.
___
*[pkg/sql/colexec/hashjoiner_tmpl.go, line 115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOo2LWl3odHg8u753ny:-MOo9F7zDYZuXdV1te4L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/sql/colexec/hashjoiner_tmpl.go#L115)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[pkg/sql/colexec/hashjoiner_tmpl.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57854#-MOo2SD3-1WzEuDBR77V:-MOoIalPBLE2p5lm7pSY:bkh35ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/603b9ae74f5daaaac062a7cb9c13fa6e0939cf78/pkg/sql/colexec/hashjoiner_tmpl.go#L124)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

These changes are too big for what is supposed to be a mechanical transformation - it's not ideal to have logic changes like this buried in such a big commit (for one it'll be hard to understand what of this diff made a problem if we turn one up later).
</blockquote></details>

I agree, extracted the changes into a separate commit. The complication here was that without logic changes BCE weren't happening.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
The benchmarks of [colexec](https://gist.github.com/yuzefovich/134bd88bc2f96a77f1fc064a9e0ab65e) and [coldata](https://gist.github.com/yuzefovich/134bd88bc2f96a77f1fc064a9e0ab65e) packages are very positive, with few regressions here and there, but none are concerning to me.

TFTR!

bors r+
